### PR TITLE
Initial Implementaion of Spritesheet/Dataset generator for the Development Workbench.

### PR DESCRIPTION
**Pull request by akhil-rana**
This works with [caMicroscope #409](https://github.com/camicroscope/caMicroscope/pull/409)

## Description
1. This helps in creation of dataset from the caMicroscope labelling files. 
2. It extracts and organises the files present in multiple zip files so that the spritesheet generation can be performed.
3. It places the images in the following fashion:
 `./workbench-utils/dataset/<labelname>/<corresponding images>` 
4. The `./workbench-utils/dataset/` folder is archived into a zip file which is returned to the user after the spritesheet and labelling files have been generated.
5. `Dockerfile` has been modified to install python and dependencies(Pillow, numpy etc.) required to run `./workbench-utils/spritemaker.py` which is responsible for spritesheet generation.
6. All the files including zips, datasets are automatically deleted when the task is done.

## Motivation and Context
- Since spritesheet generation can be a quite heavy task because of direct pixel-level processing of many images so this task has to performed on the server-side.

## How Has This Been Tested?
Briefly tested on Firefox (78.0), Chrome (84.0) on Windows 10 and Linux (arch)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: workbench-utils/spritemaker.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                    one_hot_y.append(1)
+                else:
+                    one_hot_y.append(0)
+    # build 1-hot encoded answer key
+
+    y_offset += 1
+
+
+# Save answers file (Y)
+newFile = open("./workbench-utils/dataset/labels.bin", "wb")
```
[the remark this answers is not in the document]
Comment (akhil-rana) [reply workbench-utils/spritemaker.py:81]: Everytime, the /dataset folder is deleted and recreated so new files will be created at every new instance.

**Comment by birm** [comment]
I'm testing now, but I'm still not sure how I feel about python in the caracal container. @nanli-emory thoughts?

File: workbench-utils/spritemaker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (birm) [reply workbench-utils/spritemaker.py:81]: What happens if you and I want to use the tool at the same time?

**Comment by nanli-emory** [comment]
@birm I think it depends on what your perspective for Caracal. For now, I think it is fine. But when all ML part business growth I think we can create a new container for ML core.

**Comment by birm** [comment]
I'm getting some errors when trying to use it.
I used the CMU3 sample image, and created ~ 5 "a" labels and 5 "b" labels, and tried to use that. The console returns
```
ca-back     | [Error: ENOENT: no such file or directory, access './workbench-utils/dataset.zip'] {
ca-back     |   errno: -2,
ca-back     |   code: 'ENOENT',
ca-back     |   syscall: 'access',
ca-back     |   path: './workbench-utils/dataset.zip'
ca-back     | }
ca-back     | POST /workbench/deleteDataset 200 1.116 ms - 24
ca-back     | Extracted
ca-back     | Unhandled rejection Error: File does not exist. Check to make sure the file path to your csv is correct.
ca-back     |     at /root/src/node_modules/csvtojson/v2/Converter.js:81:37
ca-back     |     at suppressedCallback (fs.js:210:5)
ca-back     |     at FSReqCallback.oncomplete (fs.js:154:23)
ca-back     | Extraction complete
```

**Comment by akhil-rana** [comment]
Okay. It is not able to locate the patches.csv file. I'm not sure why that is happening. It worked fine for me in all the cases. I might have missed something. I'm into it.

**Comment by akhil-rana** [comment]
Okay it might be checking for the csv file even before it's extraction. I'll try to solve this.

File: workbench-utils/spritemaker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (akhil-rana) [reply workbench-utils/spritemaker.py:81]: Oh, okay that is an edge case here. I overlooked that. What do you suggest. Should we make individual folders per user?

**Comment by akhil-rana** [comment]
> I'm getting some errors when trying to use it.
> I used the CMU3 sample image, and created ~ 5 "a" labels and 5 "b" labels, and tried to use that. The console returns
> 
> ```
> ca-back     | [Error: ENOENT: no such file or directory, access './workbench-utils/dataset.zip'] {
> ca-back     |   errno: -2,
> ca-back     |   code: 'ENOENT',
> ca-back     |   syscall: 'access',
> ca-back     |   path: './workbench-utils/dataset.zip'
> ca-back     | }
> ca-back     | POST /workbench/deleteDataset 200 1.116 ms - 24
> ca-back     | Extracted
> ca-back     | Unhandled rejection Error: File does not exist. Check to make sure the file path to your csv is correct.
> ca-back     |     at /root/src/node_modules/csvtojson/v2/Converter.js:81:37
> ca-back     |     at suppressedCallback (fs.js:210:5)
> ca-back     |     at FSReqCallback.oncomplete (fs.js:154:23)
> ca-back     | Extraction complete
> ```

@birm  You are getting this everytime you run the program? Or randomly? also are the files you using too big? Because I'm not able to reproduce this `unhandled rejection` error.

**Comment by birm** [comment]
I'll recommend moving python code to the slideloader service (or a new service in anticipation of more ml utilities), in any case. Let me know if you need either created.

**Comment by akhil-rana** [comment]
I don't think I'll need python for anything else other than spritesheet generation. Though I might need nodejs for more utilities in future. Do you suggest making a new nodejs container for ML services?

**Comment by akhil-rana** [comment]
Or for now I can try and move the python code to slideloader and make changes here to reflect the same. And if I need more ML services in the future I'll let you know.

**Comment by akhil-rana** [comment]
@birm 
I've implemented the spritesheet generator in [Slideloader #32](https://github.com/camicroscope/SlideLoader/pull/32).
I think we can close this now if you say so.

**Comment by birm** [comment]
Don't we need to add the workbench routes to caracal somehow, though?

**Pull request closed without merging**